core/state-machine/src/backend.rs

```diff
@@ -239,6 +249,41 @@ impl<H: Hasher> Backend<H> for InMemory<H> where H::Out: HeapSizeOf {
 		self.inner.get(&Some(storage_key.to_vec())).map(|map| map.keys().for_each(|k| f(&k)));
 	}
 
+	fn full_storage_root<I>(&self, delta: I) -> (H::Out, Self::Transaction)
```

Comment:
I think this may not give the expected result.

When we call `Backend::storage_root` / `Backend::full_storage_root`, the backend is without the overlay information. So below when we iterate over the inner child storage, we're iterating over the *last committed* values. This is not really right because the root storage provided with `delta` contains the `currently overlayed` values.